Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-47.ca-central-1.compute.internal
 Git:  git version 2.9.5
Sha: acd32f03ae2e3d0cfb0039db00e5fb529057140d
Containerized:False
Number of Error(s): 8
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
/home/centos/smoketest/PR-15242/HPCC-Platform/roxie/ccd/ccdserver.cpp:422:87: error: \xe2\x80\x98StTimeAgentWait\xe2\x80\x99 was not declared in this scope
/home/centos/smoketest/PR-15242/HPCC-Platform/roxie/ccd/ccdserver.cpp:426:69: error: no matching function for call to \xe2\x80\x98StatisticsMapping::StatisticsMapping(<brace-enclosed initializer list>)\xe2\x80\x99
make[2]: *** [roxie/ccd/CMakeFiles/ccd.dir/build.make:128: roxie/ccd/CMakeFiles/ccd.dir/ccdserver.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:16820: roxie/ccd/CMakeFiles/ccd.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2021-08-06_09-02-34
time logs started
Prep time   :     8 sec (00:00:08)
Build time  :   280 sec (00:04:40)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   288 sec (00:04:48)
time logs end

[Build error:](https://gist.github.com/de20e7382d37fde8487073a52887579e)
Build log [Build log](https://gist.github.com/HPCCSmoketest/de20e7382d37fde8487073a52887579e#file-result-21-08-06-08-57-33-log)

--- a/roxie/ccd/ccd.hpp
+++ b/roxie/ccd/ccd.hpp
@@ -552,6 +552,8 @@ public:
 extern void putStatsValue(IPropertyTree *node, const char *statName, const char *statType, unsigned __int64 val);
 extern void putStatsValue(StringBuffer &reply, const char *statName, const char *statType, unsigned __int64 val);
 
+extern const StatisticsMapping globalStatistics;
+
 class ContextLogger : implements IRoxieContextLogger, public CInterface
 {
 protected:
@@ -576,7 +578,7 @@ private:
 public:
     IMPLEMENT_IINTERFACE;
 
-    ContextLogger() : stats(allStatistics)
+    ContextLogger() : stats(globalStatistics, true)
     {
         ctxTraceLevel = traceLevel;
         intercept = false;

--- a/roxie/ccd/ccdserver.cpp
+++ b/roxie/ccd/ccdserver.cpp
@@ -406,6 +406,25 @@ static const StatisticsMapping groupStatistics({ StNumGroups, StNumGroupMax }, a
 static const StatisticsMapping sortStatistics({ StTimeSortElapsed }, actStatistics);
 static const StatisticsMapping indexWriteStatistics({ StNumDuplicateKeys }, actStatistics);
 
+// These ones get accumulated and reported in COMPLETE: line (and workunit). Excludes ones that are not sensible to sum across activities
+
+extern const StatisticsMapping globalStatistics({StWhenFirstRow, StTimeLocalExecute, StSizeMaxRowSize,
+                                                 StNumRowsProcessed, StNumSlaves, StNumStarts, StNumStops, StNumStrands,
+                                                 StNumScansPerRow, StNumAllocations, StNumAllocationScans,
+                                                 StTimeFirstExecute, StCycleLocalExecuteCycles,
+                                                 StNumAtmostTriggered,
+                                                 StNumServerCacheHits, StNumIndexSeeks, StNumIndexScans, StNumIndexWildSeeks,
+                                                 StNumIndexSkips, StNumIndexNullSkips, StNumIndexMerges, StNumIndexMergeCompares,
+                                                 StNumPreFiltered, StNumPostFiltered, StNumIndexAccepted, StNumIndexRejected,
+                                                 StNumIndexRowsRead, StNumDiskRowsRead, StNumDiskSeeks, StNumDiskAccepted,
+                                                 StNumBlobCacheHits, StNumLeafCacheHits, StNumNodeCacheHits,
+                                                 StNumBlobCacheAdds, StNumLeafCacheAdds, StNumNodeCacheAdds,
+                                                 StNumDiskRejected, StSizeAgentReply, StTimeAgentWait,
+                                                 StTimeSoapcall,
+                                                 StNumGroups,
+                                                 StTimeSortElapsed,
+                                                 StNumDuplicateKeys});
+
 //=================================================================================
 
 extern SinkMode getSinkMode(const char *val)

--- a/system/jlib/jstats.h
+++ b/system/jlib/jstats.h
@@ -510,12 +510,18 @@ class CNestedRuntimeStatisticMap;
 class jlib_decl CRuntimeStatisticCollection
 {
 public:
-    CRuntimeStatisticCollection(const StatisticsMapping & _mapping) : mapping(_mapping)
-    {
+    CRuntimeStatisticCollection(const StatisticsMapping & _mapping, bool _ignoreUnknown = false) : mapping(_mapping)
+#ifdef _DEBUG
+    ,ignoreUnknown(_ignoreUnknown)
+#endif
+{
         unsigned num = mapping.numStatistics();
         values = new CRuntimeStatistic[num+1]; // extra entry is to gather unexpected stats
     }
     CRuntimeStatisticCollection(const CRuntimeStatisticCollection & _other) : mapping(_other.mapping)
+#ifdef _DEBUG
+    , ignoreUnknown(_other.ignoreUnknown)
+#endif
     {
         unsigned num = mapping.numStatistics();
         values = new CRuntimeStatistic[num+1];
@@ -527,13 +533,19 @@ public:
     inline CRuntimeStatistic & queryStatistic(StatisticKind kind)
     {
         unsigned index = queryMapping().getIndex(kind);
-        dbgassertex(index < mapping.numStatistics());
+#ifdef _DEBUG
+        if (!ignoreUnknown)
+            dbgassertex(index < mapping.numStatistics());
+#endif
         return values[index];
     }
     inline const CRuntimeStatistic & queryStatistic(StatisticKind kind) const
     {
         unsigned index = queryMapping().getIndex(kind);
-        dbgassertex(index < mapping.numStatistics());
+#ifdef _DEBUG
+        if (!ignoreUnknown)
+            dbgassertex(index < mapping.numStatistics());
+#endif
         return values[index];
     }
 
@@ -602,6 +614,9 @@ protected:
     CRuntimeStatistic * values;
     std::atomic<CNestedRuntimeStatisticMap *> nested {nullptr};
     static CriticalSection nestlock;
+#ifdef _DEBUG
+    bool ignoreUnknown = false;
+#endif
 };
 
 //NB: Serialize and deserialize are not currently implemented.

